### DISCUSSION TITLE
Support for unmanaged DLLs

### DISCUSSION THREAD

**Discussion by AmbulantRex** [jellyfin/jellyfin#9523] [General]
## The Problem
Plugins that leverage wrapper libraries for .NET native interoperability will have unmanaged DLLs in the plugin directory. Examples of this would include things like machine learning libraries that are commonly wrappers around popular native libraries (see https://github.com/charlesw/tesseract as an example). 

Assemblies are loaded by creating a new `AssemblyLoadContext` for each plugin after which it will load each DLL in the plugin's directories with the `LoadAssemblyFromPath` method. This works for managed DLLs, but unmanaged DLLs will throw a `BadImageFormatException`. This exception would be caught and mark the plugin as "Malfunctioned" .

The relevant code for loading assemblies is:
https://github.com/jellyfin/jellyfin/blob/6a39882a70fed9f1333f7673292583c08f0904fc/Emby.Server.Implementations/Plugins/PluginManager.cs#L102-L187

## Proposal: Ignore File
Wrapper libraries will often (mostly?) use their own logic to load unmanaged DLLs. The important thing is that the assembly loader leaves them out of the list so it doesn't mistakenly mark the plugin as "Malfunctioned". I propose that there be an ignore file such as `.jellyignore` à la `.dockerignore` or `.gitignore`. It would follow the same, relatively simple syntax:

| Symbol | Purpose | 
| ------- | -------- |
| ` ` | Blank or newlines are ignored.  
| `#` | Comments |
| `*` | Matches anything except a slash. |
| `**` | Subdirectory matching. A leading `**/` matches any directory relative to the ignore file. A trailing `/**` matches everything inside the preceding path (ex: `abc/**` matches all within the `abc/` path). A `a/**/b` "middle" configuration would match only paths that end with a file or directory `/b` (ex: `a/**/b` would match `a/x/y/b` and `a/yellow/b` but not `a/notb`)  |
| `?` | Matches any one character except a slash. | 
| `/` | Directory separator. All paths relative to the ignore directory. |
| `!` | Optional negation literal |
| `\` | Escape character. Use of `\\` would be treated as a directory separator. |
| `[ ]` | Range notation (ex: `[a-zA-Z]`) allowing matching of one character specified within the range. Use of the `!` or `^` would allow negation of the range |
| `^` | Alternative range negation. Works like `!` but only within ranges. |

Similar to `.gitignore`, a `.jellyignore` in subsequent directories would define new rules relative to that directory. This would allow finer grained control over ignore. 

```yaml
# Example ignore file
x64/**
x86/**
```

## Proposal 2: List of Unmanaged DLLs
A simpler alternative could be to just allow a list of unmanaged DLLs in some other type of file that requires explicit definition of each DLL's relative path. If assemblies are the only thing this would be useful for, it could be much simpler to not worry about ignore syntax. It could also be included in some metadata yaml or json file. The downside is that it requires more verbose enumeration of the files. 

```yaml
# Simple DLL Ignore File
x64/unmanaged-x64.dll
x86/unmanaged-x86.dll
```

**Comment by nielsvanvelzen** [comment]
I'm not a huge fan of an additional file that would need to be read when the server starts. Plugins installed from a repository already have a `meta.json` file with the metadata for the installed version. We could add a new property called `"assemblies"` to that to specify which DLL files to load. Omitting the property will import all DLL files (current behavior).

Would look something like this:

```json
{
//..existing metadata
  "assemblies": [
    "Jellyfin.Plugin.CoolStuff.dll"
  ]
}
```

**Reply by AmbulantRex** [reply]
I like this. I had seen the `meta.json` file but couldn't find an example where it was being packaged with the plugin from the developer. It seemed like it was being created. If a plugin developer can supply their own, then I think this is a great solution!

**Reply by crobibero** [reply]
The `manifest.json` is what is used to generate `meta.json`, so I believe we could enhance the manifest to add `assemblies`
Note that we would likely need to also update the `build.yaml` file that many plugins utilize to generate the `manifest.json`

**Reply by AmbulantRex** [reply]
In the absence of the `assemblies` property, would we want backwards compatibility and fallback to loading all assemblies?

**Reply by crobibero** [reply]
Yes, I don't think specifying assemblies to load should be a required field

**Reply by oddstr13** [reply]
`build.yaml` is used for generating the plugin metadata file `meta.json` (which was originally introduced to aid in adding plugin releases to the plugin repo, making the plugin zip self-contained with the required metadata) and the plugin repository manifest (usually something like `manifest.json`, but could be named anything) .

This is done by [`jprm`](https://github.com/oddstr13/jellyfin-plugin-repository-manager), which would need patches for new fields.
Poke me here and on Matrix when/if someting is decided on, and I'll get it added.

---

Note to self;
Relevant section is https://github.com/oddstr13/jellyfin-plugin-repository-manager/blob/4a2b165a3dc2cfc392db865a85f23c02e9f78c68/jprm/__init__.py#L478-L496
Possibly with some extra logic here to include the mentioned dlls by default, allowing avoiding duplicating the filename in the build.yaml https://github.com/oddstr13/jellyfin-plugin-repository-manager/blob/4a2b165a3dc2cfc392db865a85f23c02e9f78c68/jprm/__init__.py#L426

**Reply by AmbulantRex** [reply]
While testing my change, I realized we probably need the assemblies defined per package version in `VersionInfo` and then get mapped to `meta.json`.  Is that the right way to go about this?

**Comment by AmbulantRex** [comment]
The following draft PR represents the change discussed: https://github.com/jellyfin/jellyfin/pull/9564

**Reply by AmbulantRex** [reply]
@oddstr13 my PR should be backwards compatible, but consider yourself poked.

**Reply by oddstr13** [reply]
Sorry for being late to the review party, but I have a question about the overall design, specifically about including the assemblies array in the plugin repo, as the info there is mainly for determining which version to install.

Would the assemblies info be needed before the server is about to load the plugin?

**Reply by AmbulantRex** [reply]
The `meta.json` was able to derive from the repo manifest. I figured it could be useful to allow it to be defined there as well. However, they'd have to define it by version. It's not technically crucial to this change. The `assemblies` array is optional and it will fallback to loading all assemblies. I could easily change it to be more opinionated like "If you want custom assemblies whitelisting, you MUST supply your own meta.json". Is this the route we want to go?

**Reply by oddstr13** [reply]
Yes, I think the `meta.json` should be a requirement, but I am biased as the creator and maintainer of JPRM.
The origin of the `meta.json` file was to serve as the metadata for adding the plugin to the plugin repository without supplying an additional file (`build.yaml`), or a bunch of command line arguments. Other parts of the project embraced it, and extended it's use cases.

Have you tested how older versions of Jellyfin behaves with new, unknown keys in the repo manifest? Do they silently discard the information, or do they error?

**Reply by AmbulantRex** [reply]
I'll need to revamp the package installation a bit then. It appears that installing forces a recreation of the meta.json file from the package and version info from the repo rather than allowing plugins to supply their own. This actually gives me some pause as they could conceivably have a mismatch between their repo manifest and the `meta.json`. 

I think we should be doing some superficial validation that the package info matches the `meta.json` and allow the rest to proceed as normal. Are there concerns about that approach?

**Reply by oddstr13** [reply]
> I'll need to revamp the package installation a bit then. It appears that installing forces a recreation of the meta.json file from the package and version info from the repo rather than allowing plugins to supply their own.

Interesting. This wouldn't be noticeable in our infrastructure, as the meta.json from the zip is used to create the repository manifest (`jprm repo add …`).

The repository *really* should be validated at repo creation/update time, not at install time.

The top level metadata (not in the version array) are not unlikely to change, and will create mismatches with the meta.json in old plugin versions. (Update of the description for example)

The UUID is the only metadata field that *must* match as far as I know. The version field really *should* match, or the update checking will misbehave (but that might be taken from the dll files)

**Reply by AmbulantRex** [reply]
> The UUID is the only metadata field that _must_ match as far as I know. The version field really _should_ match, or the update checking will misbehave (but that might be taken from the dll files)

I think I'll take a three pronged approach then:
1. The meta.json must be able to deserialize into the `PluginManifest`. This will be an issue in a few places so it makes sense to check here. This also needs to work for the other two conditions. 
2. A mismatch of UUID would be considered an error status resulting in a "Malfunctioned" state. 
3. A mismatch of `Version` will log a warning and continue. 

Regarding the first one, should that condition just result in overwriting the `meta.json` using the package info like before or should it be an exception case?

**Reply by oddstr13** [reply]
1 should accept (and preferably preserve) arbitrary additional keys, as more may be added in the future (like now).
Should probably also fill in any empty keys for which the data is available (like version), and log a warning for other mismatches.
If it cannot be deserialized into a dictionary, it should log an (non-fatal) error and then overwrite in my opinion.

2 and 3 sounds good.

@crobibero @BaronGreenback Do you guys have any opinions on this?

**Reply by AmbulantRex** [reply]
I believe JSON deserialization will simply ignore additional properties. Filling in missing fields in the local `meta.json` could leverage reflection due to the infrequency of plugin installation and anemic structure of the model.

**Reply by AmbulantRex** [reply]
> 1 should accept (and preferably preserve) arbitrary additional keys, as more may be added in the future (like now).

Upon closer inspection, the `meta.json` is routinely clobbered when changes occur. Deserializing to a dictionary could lose the intent of the `meta.json` and promote it as a sort of dumping ground for plugin settings. For that reason, I think we should stick with a known type. I'll definitely attempt to reconcile empty or default values against the package info if possible.

**Reply by crobibero** [reply]
I completely agree with 1, 2, 3

The json deserializer already ignores unknown properties, so please don't add a dictionary to carry the unknown ones.

**Reply by AmbulantRex** [reply]
My latest changes are in the above PR. It should now reflect this discussion.
